### PR TITLE
Move as much as possible into abstract Engine

### PR DESCRIPTION
**Pull request by dakrone**
This paves the way for more shared code between the `InternalEngine` and
`ShadowEngine` by way of the abstract `Engine` class. No actual
functionality has been changed.

### CODE REVIEW

File: src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 1]
```diff
@@ -181,6 +198,35 @@ public Condition newCondition() {
 
     public abstract void delete(DeleteByQuery delete) throws EngineException;
 
+    protected GetResult getFromSearcher(Get get) throws EngineException {
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/index/engine/Engine.java]: can this be final?

File: src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 2]
```diff
@@ -190,22 +236,145 @@ public Condition newCondition() {
...
-    public abstract Searcher acquireSearcher(String source) throws EngineException;
+    public final Searcher acquireSearcher(String source) throws EngineException {
+        boolean success = false;
+         /* Acquire order here is store -> manager since we need
+          * to make sure that the store is not closed before
+          * the searcher is acquired. */
+        store.incRef();
+        try {
+            final SearcherManager manager = getSearcherManager(); // can never be null
+            assert manager != null : "SearcherManager is null";
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/index/engine/Engine.java]: this can go away?

**Comment by s1monw** [comment]
left minor comments LGTM feel free to push once fixed

**Comment by dakrone** [comment]
Merged this to 1.x and master

**Pull request closed without merging**